src/tabs/index.tsx

```diff
@@ -60,6 +72,18 @@ export default function Tabs({
     changeHandler: 'onChange',
   });
 
+  const [viewedTabs, setViewedTabs] = useState(new Set(activeTabId));
+
+  useEffect(() => {
```

Comment:
This effect can be removed in favour of decorating the setActiveTabId method, e.g.

```
const setActiveTabIdDecorated = (newActiveTabId) => {
  setActiveTabId(newActiveTabId);
  setViewedTabs(prev => new Set(prev).add(newActiveTabId));
}
```

This way the related state is set within one render cycle so should be better for performance and is better clear, too.

Reply:
Given that this is a controllable component, I think we need to use an effect. Although it's unlikely, the active tab could change without `setActiveTabId` being called

Reply:
Oh, I see. Can the visited tabs be a ref then, e.g.:

```
const visitedTabsRef = useRef(new Set<string>());
visitedTabsRef.current.add(activeTabId);
```

Since changing the tab always causes a new render no matter if done in controlled or uncontrolled way, I think this approach should work, too.

Reply:
good point: there's no explicit re-render required here, and the updated value isn't needed until later (switching away from the tab) anyway